user file prefix [WIP]

capnramses opened the pull request:
## Purpose of this PR

* This is a user-requested feature for vol2obj command line tool that allows a user to specify the output filename prefix, in place of the default `output_frame_`.
* The new flag `--prefix` or `-p` followed by a string of characters, without spaces, will be used as the prefix. This can help differentiate between multiple outputs written to the same folder.
* Internal code VOLUAPP-1083.

## Changes made in this PR

Write a summary here describing **_what_** work you did in your branch, and **_how_** you did it (explain your approach).

* Added a new flag to specify prefix.
* Also made sure all the command line flags had both `-p` single and `--prefix` word options, including mandatory flags `-h` `--header`, `-s` `--sequence`, and `-v` `--video`.
* Try to attach **screenshots** wherever possible - a picture is worth 1000 words.
* An attached video, or gif, may be worth even more.
* Did you update any unit tests or automated tests to cover your changes e.g. scripts run in GitHub Actions?
* Did you update the docs to cover new interface code? This could be function interface comments, or external documentation, as project-appropriate.
* Did you update comments in code to explain difficult-to-read sections?

## Testing Summary

Provide more detail here as required.

- [x] Explained how reviewers build, deploy, and test your code and what should they look for that would indicate a bug or regression.

* Pull, `make`, `./vol2obj.bin` or `.\vol2obj.exe` and the new flag is `--prefix` or `-p`.
* Other new flags are `--header`, `--sequence`, `--video`, `-d`.
* I changed the output of the `--help` command to reflect this.

- [x] Tested locally. Which platforms? OS/hardware/devices/compilers used.

* Ubuntu + GCC 11.2.

- [x] Stated supported platforms have you not tested on, that you need help with?

* I cannot test on macOS and need a hand here.
* I still need to test on Windows.

- [x] Named specific tools have you ran over the code: e.g. which { compilers, debuggers, fuzzers, profilers, sanitisers, linters, other tools }.

* Just GCC with `-Wall -Wextra -pedantic`.
* Recompiled with Asan just now, and found the following issues:

```
tools/vol2obj/main.c:244:17: runtime error: load of misaligned address 0x7f126b4ad80d for type 'const float', which requires 4 byte alignment
0x7f126b4ad80d: note: pointer points here
 10 8e 01 00 6c 7b 6b  3f b0 8f c2 3f 53 b2 1c  3e 07 99 6c 3f ed bb c2  3f ff cd 1b 3e b6 84 6c  3f
```

* Same issue on lines 244,245,246,253,254,261,262,263,279,280.281.
* UBSan found: `UndefinedBehaviorSanitizer: undefined-behavior tools/vol2obj/main.c:281:22`
* UBSan + ASan found known issues in library stb_image_write.
* Will investigate before inviting reviewers.

- [x] Explained inputs used to test for bugs. e.g. existing known working data, large files, empty files, etc.

* Tested with rafa t-pose, russ, ulysses 8GB hd. All produced correct (and correctly named) output that loaded into Blender.
* Tested with combinations of new and old flags.
* Tested with no and malformed flags (help menu)
* Tested with no prefix to see if the default still worked.

## Risk Introduced and any Breaking Changes

- [x] Gaps in testing: Are you unable to test on some platform? Or is some input data not available to you for testing?

* Anyone testing with any additional volograms that are known to work before would be helpful.

- [x] Analysed any other potential risks here.

* I have not made any assumptions about the validity of input characters, so it may be possible to produce invalid, or hidden, files on some operating systems if characters such as `?*/\.` are used. I could check for these (and add that to this PR), but this might cause problems if input text is not ASCII but e.g. UTF-8 or UTF-16 and second or third bytes in a codepoint are interpreted as individual ASCII characters.

## Pre-Merge Checklist

* If you have a problem with the review process (e.g. it's taking too long, or the reviewers forgot) get in touch with the Team Lead of the project!

- [ ] Before making your PR, refresh your branch with a merge from its parent in case of conflicting changes since you started your branch.
- [ ] This PR description is written clearly so that reviewers can understand the why/what/how of the work done.
- [ ] Reviewed your own PR and checked your changed code.
- [ ] Invited a range of good reviewers.
- [ ] Wait for a PR to be properly reviewed. Responded to all questions, and made any required changes.
- [ ] Received one other reviewer's approval. Ideally wait for 2+ approvals, especially if you didn't get a thorough reivew on the first approval.
- [ ] If a reviewer asks a question, even if you have another approval, got the approval of the other reviewers as well before merging.
- [ ] Made sure existing reviewers approved any changes to the PR.
- [ ] Resolved any merge conflicts after making the PR.
- [ ] CI/CD builds are not reporting any errors.

patrick-volograms reviewed (approved):
Tested on MacOS with the following commands: 

`make` : 
Worked fine

`./vol2obj.bin -h` : 
Worked fine - output was very clear 

`./vol2obj.bin --header samples/cube_hdr.vol --sequence samples/cube_seq.vol --prefix hello_world --output-dir ~/Desktop/vol2obj_test` : 
Note that there is no video input (I forgot), the output was `Using output prefix = 'hello world'` followed by the output of `-h`. Although no message to tell me that I had forgotten the video input. Also note that I entered the `output_dir` flag incorrectly. 

`./vol2obj.bin --header samples/cube_hdr.vol --sequence samples/cube_seq.vol --prefix hello_world --output-dir /Users/patrick/Desktop/vol2obj_test_2 --video samples/counter.mp4` : 
Note the incorrect output dir flag. No error was thrown, but the output dir was ignored and the files were outputted to the working directory. 

`./vol2obj.bin --header samples/cube_hdr.vol --sequence samples/cube_seq.vol --prefix hello_world --output_dir /Users/patrick/Desktop/vol2obj_test_2 --video samples/counter.mp4` : 
No errors thrown. Output: 

![Screenshot 2022-06-14 at 15 46 55](https://user-images.githubusercontent.com/68862565/173607108-77444a1c-1bb2-4572-bc0b-8a8ec314ef62.png)

I think there could be some feedback to help make it fool-proof from people like me, but apart from that, it's good.

valeria-volu reviewed (approved):
Tested on Windows - all works. 
Some comments:
1) I was trying "output-dir" instead of "output_dir" and could not figure out while the save was in the exe folder. Would be safer to check for arguments that don;t exist as flags. I think Patrick already reported this.
2) Possibly remove logging for frames which are not processed  - for example, when processing frames 100 to 105, I still got 100 lines of logs for preceeding frames, like "vol_av: Frame 18 (type=P, size=14165 bytes, format=0) pts 8704 key_frame 0 [DTS 17]"
3) If would be better to check for existence of texture file before OBJs and MTLs are saved. Because currently if the file does not exist some files are still created, which is not nice

All good other than the above. I could not reproduce what Patrick was saying about not providing "-v" flag - got an error if "-h", "-v" or "-s" were not present. May be this works on Windows, but not Mac

capnramses commented:
Thanks for the reviews!

* I am going to add some parsing to warn about unknown params.
* I realised why logs were spamming - I haven't updated the libraries, so I'm doing that now (but that will mean I will need to ask for another blast of testing).

valeria-volu commented:
Sounds good @capnramses

capnramses commented:
Okay, I:

* updated vol_libs (should reduce output spam somewhat)
* tided the error message output
* replaced `--output_dir` with `--output-dir` because this seems more natural (note this is a breaking change but we can just mention it in the release notes in case someone has a script with `--output_dir` in it).
* added code to check for invalid parameters starting with `-`, but with an unrecognised option.
* added code to check for this scenario `-h -s sequencefile -v videofile` where the user forgot to put headerfile after -h, and "-s" was picked up as the header file.
* one thing i didn't handle, since it required a little extra code, was looking for unknown params where e.g. the user forgot the `-` dash before a param. i could add this, but i'd need to do more thorough validation to check if all the arguments were either a valid flag _or_ the path from a the previous valid flag. what do you think, worth doing too?

patrick-volograms commented:
Tested again after updates. Used the same commands in the same order as my previous comment.

`--help`: 
Output was `ERROR: Header argument, -h MYHEADER.vols, is mandatory.` 

`--header samples/cube_hdr.vol --sequence samples/cube_seq.vol --prefix hello_world --output-dir ~/Desktop/vol2obj_test`:
Output was `ERROR: Header argument, -h MYHEADER.vols, is mandatory.` 

`--header samples/cube_hdr.vol --sequence samples/cube_seq.vol --prefix hello_world --output-dir /Users/patrick/Desktop/vol2obj_test_2 --video samples/counter.mp4`:
No errors, everything worked correctly 

`--header samples/cube_hdr.vol --sequence samples/cube_seq.vol --prefix hello_world --output_dir /Users/patrick/Desktop/vol2obj_test_3 --video samples/counter.mp4`:
No errors, but the output_dir flag was ignored with no error, files were outputted in the working directory

capnramses commented:
Changes.

##Referring to Patrick's comment

* Fixed the new bug I introduced where `--help` stopped working properly.
* The second example is missing the video param. I'm not sure why it was warning about the header file - couldn't repro, it correctly complained about the video param for me, so maybe I fixed it since the comment but before resuming this week.
* haven't fixed the output_dir param - that's definitely a new bug i can repro and will fix in this PR.

##Other issues

* Tidied output printfs - changed makefile to not included extra debug spam from vol_libs. ( still quite spammy but its useful for debugging right now ).
* Tidied output printfs - made a function to print them by type: info/debug/warning/error/success which redirects warn/error to stderr, and colours the terminal text white/white/yellow/red/green respectively.
* Tidied punctuation in code comments.
* When an image fails to write, rather than warn, it now produces and error and stops the program. It was previously carrying-on and announcing "job done" at the end.

##Still to do

[ ] Fix `--output_dir` option.
[ ] There was another thing... my memory is poor! There is one more issue to fix though.

* Will handle out-of-disk-space issues discovered in another ticket.

capnramses commented:
I did some poking around to see how other projects handle command line args in a GNU-ish or Unixy tool fashion - seems like using a struct like I have there with flags and help text to iterate through is pretty typical. I'm not aware of any standard lib/tool/style for doing this as I don't usually write command-line tools.
I'd also be interested to know how some tools register tab auto-completion for --options flags.

capnramses commented:
![image](https://user-images.githubusercontent.com/1935602/176467914-d4b21291-8656-4ae8-8b2e-725116e91c2a.png)

capnramses commented on lib/vol_av.h:
```diff
@@ -3,7 +3,7 @@
  * vol_av    | Audio-Video Decoding API
  * --------- | ----------
- * Version   | 0.7.1
+ * Version   | 0.9
```
these files (vol_av.*, vol_geom.*) are libraries updated. ignore changes here - reviewed in vol_libs PRs.

capnramses commented on tools/vol2obj/main.c:
```diff
@@ -66,6 +66,7 @@
 #define STB_IMAGE_WRITE_IMPLEMENTATION
 #include "stb/stb_image_write.h"
 #include <assert.h>
+#include <stdarg.h>
```
i was repeating code a lot so made a printf() wrapper function that could print text in different colours. hence needing `stdarg.h`

capnramses commented on tools/vol2obj/main.c:
```diff
@@ -98,6 +125,110 @@ typedef enum img_fmt_t {
+  return false;
+}
+
+/** If command-line options are valid, their index in argv is stored here, otherwise it is 0. */
```
this array keeps track of each option flag - if it's in argv then its corresponding element here gives the argv element it appears in, otherwise 0.

capnramses commented on tools/vol2obj/main.c:
```diff
@@ -110,21 +241,7 @@ static char _output_mesh_filename[MAX_FILENAME_LEN]; // e.g. `output_frame_00000
-  }
-  return 0;
-}
+static char _prefix_str[MAX_FILENAME_LEN];           // defaults to `output_frame_`
```
the original point of the PR was just to add this one string in as a new command line option. `--prefix mysequence_`

capnramses commented on tools/vol2obj/main.c:
```diff
@@ -220,14 +340,14 @@ static bool _write_mesh_to_obj_file( const char* output_mesh_filename, const cha
   if ( !output_mesh_filename ) { return false; }
 
-  // TODO(Anton) validate fprintfs here too
+  // TODO(Anton) validate fprintfs here too for e.g. failures when running out of disk space.
```
doing this in the next PR

capnramses commented on tools/vol2obj/main.c:
```diff
@@ -522,105 +650,82 @@ int main( int argc, char** argv ) {
-        fprintf( stderr, "ERROR: -f parameter must be followed by a frame number.\n" );
-        return 1;
+    // Check for command line parameters.
+    if ( !_evaluate_params() ) { return 1; }
```
this calls the new param evaluation code

valeria-volu commented:
Re-tested. All worked as expected. Help option works. Removing required options gives error correctly. Passing options that don;t exist errors. Out of space stops processing.
